### PR TITLE
feat(gpu): block low-prio process till gpu is ready again

### PR DESCRIPTION
**Pull request by keyvank**
Do not switch to CPU when GPU is occupied!

### CODE REVIEW

File: src/gpu/locks.rs [hunk 1]
```diff
@@ -96,30 +112,32 @@ macro_rules! locked_kernel {
...
-                if let Some(ref mut k) = self.kernel {
-                    match f(k) {
-                        Err(e) => {
-                            if let GPUError::GPUTaken = e {
-                                self.free();
+                loop {
+                    if let Some(ref mut k) = self.kernel {
+                        match f(k) {
+                            Err(e) => {
+                                if let GPUError::GPUTaken = e {
```
Comment (vmx) [inline src/gpu/locks.rs]: Before even trying this out. I think you could move this `if let` directly into the matching clause, something like:

```rust
match f(k) {
    Err(GPUError::GPUTaken) => …
    Err(error) => …
    Ok(v) => …
}

And another note. I personally find `if let … else` clauses strange. Most of the time I find using matching clauses in such cases instead. I use `if let` only if there is no `else` clause.

Comment (keyvank) [reply src/gpu/locks.rs]: thanks, it looks much nicer now

**Comment by dignifiedquire** [comment]
@vmx can you test this out please, and report your results`

**Comment by vmx** [comment]
Please note: I've run all of tests only once, so expect some noise.

Proofs run for 5 mins in a loop, I count the iterations. They are run with multiple threads, not multiple processes. There is always a single high priority proof and the several low priority ones that should not have any impact on how many high priority ones are run. Here are the results.

The metrics are:

 - num low: Number of low priority threads
 - iter low: Number of iterations per low priority thread
 - iter high: Number of iterations high priority thread

| num low | iter low | iter high |
| - | -- | -- |
| 0 |  0 | 35 |
| 1 | 35 | 35 |
| 4 | 24 | 33 |
| 9 |  9 | 30 |

The reason why I also test with more than with more than 2 concurrent threads is, that due to https://github.com/filecoin-project/rust-fil-proofs/issues/1048 there are large gaps on the GPU usage, hence the low priority proof can easily run during that time. You can see the effect when I have only a single low priority thread, it manages to run as often as the high priority one as there is enough GPU time available.

Though the good news is that even with 9 low priority proofs there is still a high number of high priority proofs. I think we can't expect that 9 concurrent proofs won't have any impact on the overall performance.
 
When running the same tests on current master (without the blocking the low priority proofs from running while the high priority one runs):

| num low | iter low | iter high |
| - | -- | -- |
| 0 |  0 | 35 |
| 1 | 34 | 34 |
| 4 | 20 | 25 |
| 9 |  7 | 13 |

Here again it's clear that 2 concurrent proofs can easily be worked on next to each other, but already with 4 low priority threads, there's a huge drop in the number of high priority proofs and it's even more significant when running 9 low priority proofs in parallel.

Conclusion: This PR is a significant improvement in regards to having a high priority proof run without much disturbance from other proofs running in parallel. One area to have a closer look could be, why running 8 other parallel proofs have an impact, as they should be blocked from running.

**Comment by keyvank** [comment]
@vmx I don't think that's a good way to measure how well it performs. I think you should run a very big low-prio proof and randomly throw high prio proofs (Sparsely, because we want to have enough time to give the GPU back to the low-prio before the next high-prio happens) and measure how long these high-prio proofs take in average and then compare that with the time a high-prio proof takes when there aren't any background low-prio proofs running. Right? @dignifiedquire

**Comment by keyvank** [comment]
Though I think we can be sure this is performing better than the previous version

**Comment by vmx** [comment]
@keyvank Good points.

@dignifiedquire I'd say the current testing is good enough (it was easy, as I just needed to run them without creating much new code). But if desired I can look into improving the test code.

**Comment by dignifiedquire** [comment]
Thanks folks. I agree with @keyvank that we can still improve the testing method, but this is good enough for me to believe we have a real improvement here.

**Pull request merged**